### ISSUE TITLE
Same workspace environment in dev mode and production

### ISSUE THREAD

**Issue by jonatanschroeder** [PrairieLearn/PrairieLearn#6998]
I thought there was an issue for this, but I couldn't find it (#3139 is related but not the same). Anyway, currently in dev mode workspaces run as root, while in production they run as a non-privileged user. This can cause confusion when testing locally vs in production, since the experience is pretty different, the user's home directory is different, and root can do things that the user can't. Ideally both should be the same or as close to each other as possible.

**Comment by mwest1066** [comment]
Yes! I think that @echuber2 also has notes about how to improve this situation.

**Comment by echuber2** [comment]
We've fixed this before by adding custom scripts to workspaces to detect what the conditions are at startup and step down if necessary. For example, this is what I included with the RStudio one:

https://github.com/PrairieLearn/PrairieLearn/blob/d1c0ea737c13ee79d0c8da3526f75b54e56f2530/workspaces/rstudio/pl-gosu-helper.sh#L1-L32

And we used this to bootstrap another entrypoint script:

https://github.com/PrairieLearn/PrairieLearn/blob/d1c0ea737c13ee79d0c8da3526f75b54e56f2530/workspaces/rstudio/Dockerfile#L114

This technique was also copied over to the "desktop" workspace. But we haven't applied it to JupyterLab, where there are still issues caused by starting up as root locally. (E.g., JupyterLab steps you down to 1000 but root owns the starter file, so you can't save and submit it.)

We _could_ start up workspaces as 1001:1001 like on the server, and local developers can also configure this with content in `/PrairieLearn/config.json`:

```json
{
    "workspaceJobsDirectoryOwnerUid": 1001,
    "workspaceJobsDirectoryOwnerGid": 1001
}
```

However, I don't know if it's going to work uniformly on macOS or WSL.

There is a related issue (like what #3139 refers to) that many base images assume the user is 1000, not 1001. So really, we should be customizing all workspaces to re-ID the default account to 1001 in the image itself (not at launch time), as in the example here: https://github.com/PrairieLearn/PrairieLearn/issues/3170 Or, we could transition to using 1000:1000 on the server instead of 1001:1001, but now that would break these other adaptations we've done for the 1001 case. It would help if `workspaceOptions` in a question `info.json` could configure which user to become, with the default being 1001 as before. (I think that's better than relying on the last USER setting in the Docker itself, but maybe I'm wrong.) This would let workspace developers just say to run as 1000 instead of 1001 and skip the modification shown in https://github.com/PrairieLearn/PrairieLearn/issues/3170, while also being backwards-compatible with existing workspaces.

If we do make that change, I'm not sure how the handling of `workspaceJobsDirectoryOwnerUid`/`workspaceJobsDirectoryOwnerGid` should change.

**Comment by echuber2** [comment]
The existing defaults are defined here, in the absence of a `/PrairieLearn/config.json` file in the container:

https://github.com/PrairieLearn/PrairieLearn/blob/f05499da8f6c397d233bd8fbfb59dee2a00e0788/lib/config.js#L215-L216

**Comment by jonatanschroeder** [comment]
> We've fixed this before by adding custom scripts to workspaces to detect what the conditions are at startup and step down if necessary.

That is done at the individual workspace version, I'm proposing something in the actual workspace host.

> 
> We _could_ start up workspaces as 1001:1001 like on the server, and local developers can also configure this with content in `/PrairieLearn/config.json`:
> 
> ```json
> {
>     "workspaceJobsDirectoryOwnerUid": 1001,
>     "workspaceJobsDirectoryOwnerGid": 1001
> }
> ```

That is for file ownership, not logged in user.

**Comment by echuber2** [comment]
> That is done at the individual workspace version, I'm proposing something in the actual workspace host.

I understand what you are proposing. This is just the remedy we've taken in the past to deal with the same/related issues. As I mentioned, we could change it to always run 1001:1001 by default. Not sure if having directories owned by 1001:1001 would cause some issue on macOS mounts for some reason, but probably there is no issue with doing so.

> That is for file ownership, not logged in user.

This is directly related because having 0 as the owner prevents non-root users from being able to edit files that they _should_ be able to edit, as I mentioned. I was helping someone with exactly this issue last night. The directories could remain owned by 0:0 as long as user 1000 or 1001 has write permissions on the starter files dropped there. Furthermore, these variables are explicitly used as the startup user/group when the container is created here:

https://github.com/PrairieLearn/PrairieLearn/blob/7adbfd492ab5af2ff75f7105f34e0772663fdd67/workspace_host/interface.js#L1188

which is why I've used the `config.json` hack locally for more accurate testing in the past.

**Comment by echuber2** [comment]
Somebody asked me about this again today, so I made a post documenting how to do it (for now) using a local `config.json` file: https://github.com/PrairieLearn/PrairieLearn/discussions/8897
